src/compiler/scala/tools/nsc/javac/JavaParsers.scala

```diff
@@ -567,10 +567,49 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
 
     def varDecl(pos: Position, mods: Modifiers, tpt: Tree, name: TermName): ValDef = {
       val tpt1 = optArrayBrackets(tpt)
-      if (in.token == EQUALS && !mods.isParameter) skipTo(COMMA, SEMI)
+
+      /** Tries to detect final static literals syntactically and returns a constant type replacement */
+      def optConstantTpe(): Tree = {
+        in.nextToken()
+
+        def constantTpe(lit: Any): Tree =
+          try TypeTree(ConstantType(Constant(lit)))
+          finally in.nextToken()
+
+        if (mods.hasFlag(Flags.STATIC) && mods.isFinal) {
+          def lit(negate: Boolean): Tree =
+            if (in.lookaheadToken == SEMI)
+              in.token match {
+                case INTLIT if tpt.tpe == IntTpe => constantTpe(in.intVal(negate).toInt)
+                case LONGLIT if tpt.tpe == LongTpe => constantTpe(in.intVal(negate))
+                case FLOATLIT if tpt.tpe == FloatTpe => constantTpe(in.floatVal(negate).toFloat)
+                case DOUBLELIT if tpt.tpe == DoubleTpe => constantTpe(in.floatVal(negate))
+                case STRINGLIT =>
+                  tpt match {
+                    case Ident(TypeName("String")) => constantTpe(in.name.toString)
```

Comment:
To pass javac, I think it could only be `public static final Object s = "hi"` which would be narrowed to a constant String from the view of Scala without the check. This would probably be ok in most cases but those where overloads or implicits would be picked differently for Object vs String. So, I guess it makes sense to keep it somewhat conservative.

Wrt qualified reference, yes, right, I guess we should support them. There's only one possible way to qualify them in Java which would be `java.lang.String`, right? Shuldn't be hard to support that here as well.